flow/TLSConfig.actor.cpp

```diff
@@ -625,196 +634,364 @@ TLSPolicy::Rule::Rule(std::string input) {
-		reason = "Cert subject error";
-		goto err;
-	}
-	for (auto& pair : verify->subject_criteria) {
-		if (!match_criteria(
-		        cert, subject, pair.first, pair.second.criteria, pair.second.match_type, pair.second.location)) {
-			reason = "Cert subject match failure";
-			goto err;
-		}
+bool PeerVerifier::matchCriteria(const X509* cert, X509_NAME* name, const NID nid, const Criteria& criteria) {
+	bool result = false;
+
+	switch (criteria.location) {
+	case X509Location::NAME:
+		m_verifyState.emplace_back("Name"sv);
+		result = matchNameCriteria(name, nid, criteria.criteria, criteria.match_type);
+		break;
+
+	case X509Location::EXTENSION:
+		m_verifyState.emplace_back("Extension"sv);
+		m_currentExtension = criteria.criteria;
+		result = matchExtensionCriteria(cert, nid, criteria.criteria, criteria.match_type);
+		break;
+
+	default:
+		// Base on the enum this should NEVER happen
+		m_verifyState.emplace_back("UnsupportedX509LocationValue"sv);
 	}
 
-	// Check issuer criteria.
-	if ((issuer = X509_get_issuer_name(cert)) == nullptr) {
-		reason = "Cert issuer error";
-		goto err;
+	if (result) {
+		m_currentName.clear();
+		m_currentExtension.clear();
+
+		m_verifyState.pop_back();
 	}
-	for (auto& pair : verify->issuer_criteria) {
-		if (!match_criteria(
-		        cert, issuer, pair.first, pair.second.criteria, pair.second.match_type, pair.second.location)) {
-			reason = "Cert issuer match failure";
-			goto err;
+	return result;
+}
+
+bool PeerVerifier::verifyCriterias(const X509* cert, X509_NAME* name, const TLSPolicy::Rule::CriteriaMap& criterias) {
+	for (const auto& [nid, criteria] : criterias) {
+		if (!matchCriteria(cert, name, nid, criteria)) {
+			return false;
 		}
 	}
+	return true;
+}
 
-	// Check root criteria - this is the subject of the final certificate in the stack.
-	cert = sk_X509_value(X509_STORE_CTX_get0_chain(store_ctx), sk_X509_num(X509_STORE_CTX_get0_chain(store_ctx)) - 1);
-	if ((subject = X509_get_subject_name(cert)) == nullptr) {
-		reason = "Root subject error";
-		goto err;
+bool PeerVerifier::verifyRule(const TLSPolicy::Rule& rule) {
+	m_verifyState.emplace_back("Rule"sv);
+
+	{
+		m_verifyState.emplace_back("Cert");
+
+		const X509* cert = sk_X509_value(X509_STORE_CTX_get0_chain(m_storeCtx), 0);
+		ASSERT(cert != nullptr);
+
+		m_verifyState.emplace_back("Subject"sv);
+		if (auto subject = X509_get_subject_name(cert); subject != nullptr) {
+			if (!verifyCriterias(cert, subject, rule.subject_criteria)) {
+				return false;
+			}
+		} else {
+			m_verifyState.emplace_back("Missing"sv);
+			return false;
+		}
+		m_verifyState.pop_back();
+
+		m_verifyState.emplace_back("Issuer");
+		if (auto issuer = X509_get_issuer_name(cert); issuer != nullptr) {
+			if (!verifyCriterias(cert, issuer, rule.issuer_criteria)) {
+				return false;
+			}
+		} else {
+			m_verifyState.emplace_back("Missing"sv);
+			return false;
+		}
+		m_verifyState.pop_back();
+
+		m_verifyState.pop_back();
```

Comment:
What are the rules for these `emplace_back` and `pop_back` calls?

Reply:
Basically when a test starts, the `m_verifyState` pushes back the name of the test, if it false, the name of the test will not pop back so the failed step is kept